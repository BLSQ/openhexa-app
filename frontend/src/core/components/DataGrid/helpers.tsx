import * as React from "react";
import { Cell } from "react-table";

export const CellContext = React.createContext<Cell | null>(null);

export function useCellContext<V = any>() {
  const cell = React.useContext(CellContext);
  if (!cell) {
    throw new Error("useCellContext muse be used inside a CellContext wrapper");
  }
  return cell as Cell<{}, V>;
}

export function CellContextProvider(props: {
  cell: Cell;
  children: React.ReactNode;
}) {
  return (
    <CellContext.Provider value={props.cell}>
      {props.children}
    </CellContext.Provider>
  );
}
