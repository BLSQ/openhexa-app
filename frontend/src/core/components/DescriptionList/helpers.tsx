import { useContext, createContext } from "react";

export enum DescriptionListDisplayMode {
  LABEL_ABOVE = "above",
  LABEL_LEFT = "left",
}

export type DescriptionListContext = {
  displayMode: DescriptionListDisplayMode;
  compact: boolean;
};

export const ctx = createContext<DescriptionListContext | null>(null);

export const useDescriptionList = () => {
  const config = useContext(ctx);

  if (!config) {
    throw new Error(
      "useDescriptionList must be under a DescriptionList component",
    );
  }
  return config;
};
