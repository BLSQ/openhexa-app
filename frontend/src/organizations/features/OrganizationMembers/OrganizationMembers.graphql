query OrganizationMembers($id: UUID!, $page: Int, $perPage: Int, $term: String) {
  organization(id: $id) {
    id
      permissions {
          manageMembers
      }
    members(page: $page, perPage: $perPage, term: $term) {
        totalItems
      items {
        id
        role
        workspaceMemberships {
            ...WorkspaceRole
            id
            role
            workspace {
                slug
                name
            }
        }
        user {
            ...User_user
        }
        createdAt
      }
    }
  }
}

mutation UpdateOrganizationMember($input: UpdateOrganizationMemberInput!) {
        updateOrganizationMember(input: $input) {
          success
          errors
          membership {
            id
            role
          }
        }
}

mutation AddOrganizationMember($input: AddOrganizationMemberInput!) {
  addOrganizationMember(input: $input) {
    success
    errors
    membership {
      id
      user {
        id
        displayName
        email
      }
      role
    }
  }
}

mutation DeleteOrganizationMember($input: DeleteOrganizationMemberInput!) {
  deleteOrganizationMember(input: $input) {
    success
    errors
  }
}

mutation InviteOrganizationMember($input: InviteOrganizationMemberInput!) {
  inviteOrganizationMember(input: $input) {
    success
    errors
    membership {
      id
      user {
        id
        displayName
        email
      }
      role
    }
  }
}
