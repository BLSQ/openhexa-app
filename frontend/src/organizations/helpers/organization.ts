import { OrganizationMembershipRole } from "graphql/types";
import { i18n } from "next-i18next";

export function formatOrganizationMembershipRole(
  role: OrganizationMembershipRole,
) {
  switch (role) {
    case OrganizationMembershipRole.Owner:
      return i18n!.t("Owner");
    case OrganizationMembershipRole.Admin:
      return i18n!.t("Admin");
    case OrganizationMembershipRole.Member:
      return i18n!.t("Member");
    default:
      return i18n!.t("Unknown");
  }
}
