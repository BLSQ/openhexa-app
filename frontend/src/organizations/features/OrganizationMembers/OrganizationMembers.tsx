import { PencilIcon, TrashIcon } from "@heroicons/react/24/outline";
import Button from "core/components/Button";
import DataGrid, { BaseColumn } from "core/components/DataGrid";
import DateColumn from "core/components/DataGrid/DateColumn";
import useCacheKey from "core/hooks/useCacheKey";
import useDebounce from "core/hooks/useDebounce";
import SearchInput from "core/features/SearchInput";
import { User as UserType, OrganizationMembership } from "graphql/types";
import { DateTime } from "luxon";
import { useState, useEffect } from "react";
import { useTranslation } from "next-i18next";
import DeleteOrganizationMemberDialog from "./DeleteOrganizationMemberDialog";
import UpdateOrganizationMemberDialog from "./UpdateOrganizationMemberDialog";
import useMe from "identity/hooks/useMe";
import {
  useOrganizationMembersQuery,
  OrganizationMembersQuery,
} from "./OrganizationMembers.generated";
import Block from "core/components/Block";
import User from "core/features/User";
import OrganizationRoleBadge from "organizations/components/OrganizationRoleBadge";
import WorkspaceRolesList from "organizations/components/WorkspaceRolesList";

const DEFAULT_PAGE_SIZE = 10;

type OrganizationMember = Pick<
  OrganizationMembership,
  "id" | "role" | "workspaceMemberships"
> & {
  user: Pick<UserType, "id" | "displayName" | "email">;
};

export default function OrganizationMembers({
  organizationId,
}: {
  organizationId: string;
}) {
  const me = useMe();
  const { t } = useTranslation();
  const [selectedMember, setSelectedMember] = useState<OrganizationMember>();
  const [openDeleteDialog, setOpenDeleteDialog] = useState(false);
  const [openEditDialog, setOpenEditDialog] = useState(false);
  const [searchTerm, setSearchTerm] = useState("");
  const [previousData, setPreviousData] =
    useState<OrganizationMembersQuery | null>(null);

  const debouncedSearchTerm = useDebounce(searchTerm, 300);

  const { data, refetch, loading } = useOrganizationMembersQuery({
    variables: {
      id: organizationId,
      page: 1,
      perPage: DEFAULT_PAGE_SIZE,
      term: debouncedSearchTerm || undefined,
    },
    fetchPolicy: "cache-and-network",
  });

  useEffect(() => {
    if (data && !loading) {
      setPreviousData(data);
    }
  }, [data, loading]);

  useCacheKey("organization", () => refetch());

  const onChangePage = ({ page }: { page: number }) => {
    refetch({
      page,
      id: organizationId,
      term: debouncedSearchTerm || undefined,
    }).then();
  };

  const displayData = data || previousData;
  const organization = displayData?.organization ?? {
    members: { items: [], totalItems: 0 },
    permissions: { manageMembers: false },
  };

  const handleDeleteClicked = (member: OrganizationMember) => {
    setSelectedMember(member);
    setOpenDeleteDialog(true);
  };

  const handleUpdateClicked = (member: OrganizationMember) => {
    setSelectedMember(member);
    setOpenEditDialog(true);
  };

  return (
    <>
      <div className="mb-4">
        <SearchInput
          value={searchTerm}
          onChange={(e) => setSearchTerm(e.target.value)}
          placeholder={t("Search members...")}
          loading={loading}
          className="max-w-md"
          fitWidth
        />
      </div>
      <Block>
        <DataGrid
          defaultPageSize={DEFAULT_PAGE_SIZE}
          totalItems={organization.members.totalItems}
          data={organization.members.items}
          fetchData={onChangePage}
          className="min-h-30"
        >
          <BaseColumn label={t("User")} id="user" minWidth={200}>
            {(membership) => <User user={membership.user} subtext />}
          </BaseColumn>
          <BaseColumn label={t("Organization Role")} id="org_role">
            {(member) => <OrganizationRoleBadge role={member.role} size="sm" />}
          </BaseColumn>
          <BaseColumn
            label={t("Workspace Roles")}
            id="workspace_roles"
            minWidth={300}
          >
            {(member) => (
              <WorkspaceRolesList
                workspaceMemberships={member.workspaceMemberships}
                size="sm"
                maxVisible={2}
              />
            )}
          </BaseColumn>
          <DateColumn
            className="py-4"
            accessor="createdAt"
            id="createdAt"
            label={t("Joined")}
            format={DateTime.DATE_FULL}
          />
          <BaseColumn className="flex justify-end gap-x-2">
            {(member) =>
              organization.permissions.manageMembers ? (
                <>
                  <Button
                    onClick={() => handleUpdateClicked(member)}
                    size="sm"
                    variant="secondary"
                  >
                    <PencilIcon className="h-4" />
                  </Button>
                  <Button
                    onClick={() => handleDeleteClicked(member)}
                    size="sm"
                    variant="secondary"
                  >
                    <TrashIcon className="h-4" />
                  </Button>
                </>
              ) : (
                <></>
              )
            }
          </BaseColumn>
        </DataGrid>
      </Block>
      {selectedMember && (
        <DeleteOrganizationMemberDialog
          open={openDeleteDialog}
          onClose={() => {
            setSelectedMember(undefined);
            setOpenDeleteDialog(false);
          }}
          member={selectedMember}
        />
      )}
      {selectedMember && (
        <UpdateOrganizationMemberDialog
          open={openEditDialog}
          onClose={() => {
            setSelectedMember(undefined);
            setOpenEditDialog(false);
          }}
          member={selectedMember}
        />
      )}
    </>
  );
}
