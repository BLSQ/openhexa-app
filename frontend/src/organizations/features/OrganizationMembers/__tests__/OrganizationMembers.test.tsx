import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { useOrganizationMembersQuery } from "../OrganizationMembers.generated";
import { TestApp } from "core/helpers/testutils";
import OrganizationMembers from "../OrganizationMembers";
import { v4 } from "uuid";
import {
  OrganizationMembershipRole,
  WorkspaceMembershipRole,
} from "graphql/types";

const ORGANIZATION_ID = v4();

const MOCK_MEMBERS = [
  {
    id: v4(),
    role: OrganizationMembershipRole.Admin,
    createdAt: "2023-01-15T10:00:00Z",
    workspaceMemberships: [
      {
        id: v4(),
        role: WorkspaceMembershipRole.Editor,
        workspace: {
          slug: "workspace-1",
          name: "Test Workspace 1",
        },
      },
    ],
    user: {
      id: v4(),
      displayName: "John Doe",
      email: "john.doe@example.com",
      avatar: {
        initials: "JD",
        color: "blue",
      },
    },
  },
  {
    id: v4(),
    role: OrganizationMembershipRole.Member,
    createdAt: "2023-02-20T14:30:00Z",
    workspaceMemberships: [
      {
        id: v4(),
        role: WorkspaceMembershipRole.Viewer,
        workspace: {
          slug: "workspace-2",
          name: "Test Workspace 2",
        },
      },
    ],
    user: {
      id: v4(),
      displayName: "Jane Smith",
      email: "jane.smith@example.com",
      avatar: {
        initials: "JS",
        color: "green",
      },
    },
  },
];

const MOCK_ORGANIZATION_QUERY_RESULT = {
  organization: {
    id: ORGANIZATION_ID,
    permissions: {
      manageMembers: true,
    },
    members: {
      totalItems: 2,
      items: MOCK_MEMBERS,
    },
    workspaces: {
      items: [
        {
          slug: "workspace-1",
          name: "Test Workspace 1",
        },
        {
          slug: "workspace-2",
          name: "Test Workspace 2",
        },
      ],
    },
  },
};

const MOCK_ORGANIZATION_QUERY_NO_PERMISSIONS = {
  organization: {
    id: ORGANIZATION_ID,
    permissions: {
      manageMembers: false,
    },
    members: {
      totalItems: 2,
      items: MOCK_MEMBERS,
    },
    workspaces: {
      items: [
        {
          slug: "workspace-1",
          name: "Test Workspace 1",
        },
        {
          slug: "workspace-2",
          name: "Test Workspace 2",
        },
      ],
    },
  },
};

jest.mock("../OrganizationMembers.generated", () => ({
  ...jest.requireActual("../OrganizationMembers.generated"),
  useOrganizationMembersQuery: jest.fn(),
}));

jest.mock("next-i18next", () => ({
  useTranslation: () => ({
    t: (key: string) => key,
  }),
  Trans: ({ children }: { children: React.ReactNode }) => <>{children}</>,
  i18n: {
    t: (key: string) => key,
  },
}));

const useOrganizationMembersQueryMock =
  useOrganizationMembersQuery as jest.Mock;

describe("OrganizationMembers", () => {
  beforeEach(() => {
    useOrganizationMembersQueryMock.mockClear();
  });

  it("renders organization members list correctly", async () => {
    useOrganizationMembersQueryMock.mockReturnValue({
      data: MOCK_ORGANIZATION_QUERY_RESULT,
      loading: false,
      refetch: jest.fn(),
    });

    const { container } = render(
      <OrganizationMembers organizationId={ORGANIZATION_ID} />,
    );

    expect(screen.getByText("John Doe")).toBeInTheDocument();
    expect(screen.getByText("jane.smith@example.com")).toBeInTheDocument();
    expect(screen.getByText("Jane Smith")).toBeInTheDocument();

    expect(screen.getByText("Admin")).toBeInTheDocument();
    expect(screen.getByText("Member")).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it("handles empty members list", async () => {
    useOrganizationMembersQueryMock.mockReturnValue({
      data: {
        organization: {
          id: ORGANIZATION_ID,
          permissions: { manageMembers: true },
          members: { totalItems: 0, items: [] },
        },
      },
      loading: false,
      refetch: jest.fn(),
    });

    render(<OrganizationMembers organizationId={ORGANIZATION_ID} />);

    expect(screen.getByText("User")).toBeInTheDocument();
    expect(screen.getByText("Organization Role")).toBeInTheDocument();
  });

  it("handles search functionality", async () => {
    let lastQueryVariables: any = {};
    const mockRefetch = jest.fn();

    useOrganizationMembersQueryMock.mockImplementation((options) => {
      lastQueryVariables = options.variables;
      return {
        data: MOCK_ORGANIZATION_QUERY_RESULT,
        loading: false,
        refetch: mockRefetch,
      };
    });

    const user = userEvent.setup();

    render(<OrganizationMembers organizationId={ORGANIZATION_ID} />);

    const searchInput = screen.getByPlaceholderText("Search members...");

    expect(lastQueryVariables).toEqual({
      id: ORGANIZATION_ID,
      page: 1,
      perPage: 10,
      term: "",
    });

    await user.type(searchInput, "John");

    await waitFor(
      () => {
        expect(lastQueryVariables).toEqual({
          id: ORGANIZATION_ID,
          page: 1,
          perPage: 10,
          term: "John",
        });
      },
      { timeout: 1000 },
    );

    expect(searchInput).toHaveValue("John");
  });

  it("shows edit and delete buttons when user has manage permissions", async () => {
    useOrganizationMembersQueryMock.mockReturnValue({
      data: MOCK_ORGANIZATION_QUERY_RESULT,
      loading: false,
      refetch: jest.fn(),
    });

    render(<OrganizationMembers organizationId={ORGANIZATION_ID} />);

    const editButtons = screen.getAllByLabelText("edit");
    const deleteButtons = screen.getAllByLabelText("delete");

    expect(editButtons).toHaveLength(2);
    expect(deleteButtons).toHaveLength(2);
  });

  it("hides edit and delete buttons when user lacks manage permissions", async () => {
    useOrganizationMembersQueryMock.mockReturnValue({
      data: MOCK_ORGANIZATION_QUERY_NO_PERMISSIONS,
      loading: false,
      refetch: jest.fn(),
    });

    render(<OrganizationMembers organizationId={ORGANIZATION_ID} />);

    const buttons = screen.queryAllByRole("button");
    expect(buttons).toHaveLength(0);
  });

  it("opens delete dialog when delete button is clicked", async () => {
    useOrganizationMembersQueryMock.mockReturnValue({
      data: MOCK_ORGANIZATION_QUERY_RESULT,
      loading: false,
      refetch: jest.fn(),
    });

    const user = userEvent.setup();

    render(
      <TestApp>
        <OrganizationMembers organizationId={ORGANIZATION_ID} />
      </TestApp>,
    );

    const deleteButton = screen.getAllByLabelText("delete")[0];
    await user.click(deleteButton);

    await waitFor(
      () => {
        expect(screen.getByRole("dialog")).toBeInTheDocument();
        expect(screen.getByText("Remove Member")).toBeInTheDocument();
      },
      { timeout: 3000 },
    );
  });

  it("opens update dialog when edit button is clicked", async () => {
    useOrganizationMembersQueryMock.mockReturnValue({
      data: MOCK_ORGANIZATION_QUERY_RESULT,
      loading: false,
      refetch: jest.fn(),
    });

    const user = userEvent.setup();

    render(
      <TestApp>
        <OrganizationMembers organizationId={ORGANIZATION_ID} />
      </TestApp>,
    );

    const editButton = screen.getAllByLabelText("edit")[0];
    await user.click(editButton);

    await waitFor(
      () => {
        expect(screen.getByRole("dialog")).toBeInTheDocument();
        expect(
          screen.getByText("Update Member Permissions"),
        ).toBeInTheDocument();
      },
      { timeout: 3000 },
    );
  });

  it("displays workspace roles correctly", async () => {
    useOrganizationMembersQueryMock.mockReturnValue({
      data: MOCK_ORGANIZATION_QUERY_RESULT,
      loading: false,
      refetch: jest.fn(),
    });

    render(
      <TestApp>
        <OrganizationMembers organizationId={ORGANIZATION_ID} />
      </TestApp>,
    );

    expect(
      screen.getByText((content) => {
        return content.includes("Test Workspace 1");
      }),
    ).toBeInTheDocument();
    expect(
      screen.getByText((content) => {
        return content.includes("Test Workspace 2");
      }),
    ).toBeInTheDocument();
  });
});
