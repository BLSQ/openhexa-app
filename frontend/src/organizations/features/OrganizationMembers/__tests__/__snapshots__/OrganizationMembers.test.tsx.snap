// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`OrganizationMembers renders organization members list correctly 1`] = `
<div>
  <div
    class="mb-4 flex gap-4"
  >
    <div
      class="group relative w-fit"
    >
      <input
        aria-describedby="search-members-description"
        aria-invalid="false"
        autocomplete="off"
        class="form-input appearance-none relative block px-3 py-2 border rounded-md focus:outline-hidden focus:z-0 sm:text-sm disabled:opacity-50 disabled:cursor-not-allowed border-gray-300 placeholder-gray-500 text-gray-900 focus:ring-blue-500 focus:border-blue-500 pr-4 pl-10 max-w-md"
        data-testid="search-input"
        id="search-members"
        name="search-members"
        placeholder="Search members..."
        type="text"
        value=""
      />
      <div
        class="absolute inset-y-0 left-0 z-0 inline-flex items-center justify-center pl-2.5"
      >
        <svg
          aria-hidden="true"
          class="h-5 text-gray-500"
          data-slot="icon"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m21 21-5.197-5.197m0 0A7.5 7.5 0 1 0 5.196 5.196a7.5 7.5 0 0 0 10.607 10.607Z"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </div>
      <div
        class="absolute inset-y-0 right-0 z-0 inline-flex items-center justify-center pr-2.5 invisible"
      />
    </div>
    <div
      class="w-30"
    >
      <div
        data-headlessui-state=""
      >
        <button
          aria-expanded="false"
          aria-haspopup="listbox"
          class="form-input relative w-full  cursor-default rounded-md bg-white px-3 py-2 pr-10 text-left disabled:cursor-not-allowed disabled:opacity-50 sm:text-sm focus:ring-blue-500 text-gray-500 border-gray-300 hover:border-gray-400"
          data-headlessui-state=""
          id="headlessui-listbox-button-b608a8f9-e22c-462e-95cb-e9c54e6fa91f"
          type="button"
        >
          <span
            class="block truncate"
          >
            All roles
          </span>
          <span
            class="pointer-events-none absolute inset-y-0 right-0 flex items-center pr-2 text-gray-400 hover:text-gray-600"
          >
            <svg
              aria-hidden="true"
              class="h-5 w-5"
              data-slot="icon"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8.25 15 12 18.75 15.75 15m-7.5-6L12 5.25 15.75 9"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
  <article
    class="sm:rounded-lg overflow-hidden shadow-xs border-b border-gray-200 bg-white"
  >
    <div
      class="relative min-h-30"
    >
      <div
        class="relative"
      >
        <div
          class="h-full overflow-x-auto"
        >
          <div
            class="pointer-events-none absolute bottom-0 left-0 top-0 bg-linear-to-r to-transparent transition-opacity duration-100 from-white w-12 opacity-0"
          />
          <div
            id="child"
          >
            <table
              class="w-full divide-y divide-gray-200  table-fixed"
              role="table"
              style="min-width: 740px;"
            >
              <thead>
                <tr
                  class=""
                  role="row"
                  style="display: flex; flex: 1 0 auto; min-width: 740px;"
                >
                  <th
                    class="bg-gray-50 text-left text-xs font-medium text-gray-500 uppercase tracking-wider px-4 py-5 first:pl-6 last:pr-6 whitespace-nowrap"
                    colspan="1"
                    role="columnheader"
                    scope="col"
                    style="box-sizing: border-box; flex: 350 0 auto; min-width: 350px; width: 350px;"
                  >
                    User
                  </th>
                  <th
                    class="bg-gray-50 text-left text-xs font-medium text-gray-500 uppercase tracking-wider px-4 py-5 first:pl-6 last:pr-6 whitespace-nowrap"
                    colspan="1"
                    role="columnheader"
                    scope="col"
                    style="box-sizing: border-box; flex: 150 0 auto; min-width: 30px; width: 150px;"
                  >
                    Organization Role
                  </th>
                  <th
                    class="bg-gray-50 text-left text-xs font-medium text-gray-500 uppercase tracking-wider px-4 py-5 first:pl-6 last:pr-6 whitespace-nowrap"
                    colspan="1"
                    role="columnheader"
                    scope="col"
                    style="box-sizing: border-box; flex: 300 0 auto; min-width: 300px; width: 300px;"
                  >
                    Workspace Roles
                  </th>
                  <th
                    class="bg-gray-50 text-left text-xs font-medium text-gray-500 uppercase tracking-wider px-4 py-5 first:pl-6 last:pr-6 whitespace-nowrap"
                    colspan="1"
                    role="columnheader"
                    scope="col"
                    style="box-sizing: border-box; flex: 150 0 auto; min-width: 30px; width: 150px;"
                  >
                    Joined
                  </th>
                  <th
                    class="bg-gray-50 text-left text-xs font-medium text-gray-500 uppercase tracking-wider px-4 py-5 first:pl-6 last:pr-6 whitespace-nowrap"
                    colspan="1"
                    role="columnheader"
                    scope="col"
                    style="box-sizing: border-box; flex: 150 0 auto; min-width: 30px; width: 150px;"
                  />
                </tr>
              </thead>
              <tbody
                class="divide-y divide-gray-200"
                role="rowgroup"
              >
                <tr
                  role="row"
                  style="display: flex; flex: 1 0 auto; min-width: 740px;"
                >
                  <td
                    class="text-sm text-gray-500 h-full px-4 py-3 first:pl-6 last:pr-6 whitespace-nowrap"
                    role="cell"
                    style="box-sizing: border-box; flex: 350 0 auto; min-width: 350px; width: 350px;"
                  >
                    <div
                      class="flex items-center gap-2"
                    >
                      <span
                        class="inline-flex shrink-0 items-center justify-center rounded-full bg-gray-500 text-sm h-8 w-8"
                        style="background: blue;"
                        title="John Doe"
                      >
                        <span
                          class="font-medium leading-none text-white"
                        >
                          JD
                        </span>
                      </span>
                      <div
                        class="truncate"
                      >
                        <div
                          class="text-inherit"
                        >
                          John Doe
                        </div>
                        <div
                          class="text-xs text-gray-400"
                        >
                          john.doe@example.com
                        </div>
                      </div>
                    </div>
                  </td>
                  <td
                    class="text-sm text-gray-500 h-full px-4 py-3 first:pl-6 last:pr-6 whitespace-nowrap"
                    role="cell"
                    style="box-sizing: border-box; flex: 150 0 auto; min-width: 30px; width: 150px;"
                  >
                    <span
                      class="inline-flex items-center truncate whitespace-nowrap rounded-md font-medium ring-1 ring-inset px-2.5 py-1 text-sm text-xs bg-orange-50 text-orange-800 ring-orange-600/20"
                    >
                      Admin
                    </span>
                  </td>
                  <td
                    class="text-sm text-gray-500 h-full px-4 py-3 first:pl-6 last:pr-6 whitespace-nowrap"
                    role="cell"
                    style="box-sizing: border-box; flex: 300 0 auto; min-width: 300px; width: 300px;"
                  >
                    <div
                      class="flex flex-wrap gap-1"
                    >
                      <span
                        class="inline-flex items-center truncate whitespace-nowrap rounded-md font-medium ring-1 ring-inset px-2.5 py-1 text-sm text-xs bg-green-50 text-green-800 ring-green-600/20"
                      >
                        Test Workspace 1 Â· Editor
                      </span>
                    </div>
                  </td>
                  <td
                    class="text-sm text-gray-500 h-full py-4 px-4 py-3 first:pl-6 last:pr-6 whitespace-nowrap"
                    role="cell"
                    style="box-sizing: border-box; flex: 150 0 auto; min-width: 30px; width: 150px;"
                  >
                    <time
                      class="whitespace-nowrap truncate"
                      datetime="2023-01-15T11:00:00.000+01:00"
                      title="2023-01-15T11:00:00.000+01:00"
                    >
                      January 15, 2023
                    </time>
                  </td>
                  <td
                    class="text-sm text-gray-500 h-full flex justify-end gap-x-2 px-4 py-3 first:pl-6 last:pr-6 whitespace-nowrap"
                    role="cell"
                    style="box-sizing: border-box; flex: 150 0 auto; min-width: 30px; width: 150px;"
                  >
                    <button
                      aria-label="edit"
                      class="relative h-fit
    inline-flex items-center justify-center transition-all
    border font-medium
    disabled:opacity-50 disabled:cursor-not-allowed
   rounded-sm focus:outline-hidden focus:ring-2 focus:ring-offset-2 text-xs px-2 py-1 shadow-xs border-transparent text-indigo-700 bg-indigo-100 hover:bg-indigo-200 focus:ring-indigo-500"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="h-4"
                        data-slot="icon"
                        fill="none"
                        stroke="currentColor"
                        stroke-width="1.5"
                        viewBox="0 0 24 24"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="m16.862 4.487 1.687-1.688a1.875 1.875 0 1 1 2.652 2.652L6.832 19.82a4.5 4.5 0 0 1-1.897 1.13l-2.685.8.8-2.685a4.5 4.5 0 0 1 1.13-1.897L16.863 4.487Zm0 0L19.5 7.125"
                          stroke-linecap="round"
                          stroke-linejoin="round"
                        />
                      </svg>
                    </button>
                    <button
                      aria-label="delete"
                      class="relative h-fit
    inline-flex items-center justify-center transition-all
    border font-medium
    disabled:opacity-50 disabled:cursor-not-allowed
   rounded-sm focus:outline-hidden focus:ring-2 focus:ring-offset-2 text-xs px-2 py-1 shadow-xs border-transparent text-indigo-700 bg-indigo-100 hover:bg-indigo-200 focus:ring-indigo-500"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="h-4"
                        data-slot="icon"
                        fill="none"
                        stroke="currentColor"
                        stroke-width="1.5"
                        viewBox="0 0 24 24"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="m14.74 9-.346 9m-4.788 0L9.26 9m9.968-3.21c.342.052.682.107 1.022.166m-1.022-.165L18.16 19.673a2.25 2.25 0 0 1-2.244 2.077H8.084a2.25 2.25 0 0 1-2.244-2.077L4.772 5.79m14.456 0a48.108 48.108 0 0 0-3.478-.397m-12 .562c.34-.059.68-.114 1.022-.165m0 0a48.11 48.11 0 0 1 3.478-.397m7.5 0v-.916c0-1.18-.91-2.164-2.09-2.201a51.964 51.964 0 0 0-3.32 0c-1.18.037-2.09 1.022-2.09 2.201v.916m7.5 0a48.667 48.667 0 0 0-7.5 0"
                          stroke-linecap="round"
                          stroke-linejoin="round"
                        />
                      </svg>
                    </button>
                  </td>
                </tr>
                <tr
                  role="row"
                  style="display: flex; flex: 1 0 auto; min-width: 740px;"
                >
                  <td
                    class="text-sm text-gray-500 h-full px-4 py-3 first:pl-6 last:pr-6 whitespace-nowrap"
                    role="cell"
                    style="box-sizing: border-box; flex: 350 0 auto; min-width: 350px; width: 350px;"
                  >
                    <div
                      class="flex items-center gap-2"
                    >
                      <span
                        class="inline-flex shrink-0 items-center justify-center rounded-full bg-gray-500 text-sm h-8 w-8"
                        style="background: green;"
                        title="Jane Smith"
                      >
                        <span
                          class="font-medium leading-none text-white"
                        >
                          JS
                        </span>
                      </span>
                      <div
                        class="truncate"
                      >
                        <div
                          class="text-inherit"
                        >
                          Jane Smith
                        </div>
                        <div
                          class="text-xs text-gray-400"
                        >
                          jane.smith@example.com
                        </div>
                      </div>
                    </div>
                  </td>
                  <td
                    class="text-sm text-gray-500 h-full px-4 py-3 first:pl-6 last:pr-6 whitespace-nowrap"
                    role="cell"
                    style="box-sizing: border-box; flex: 150 0 auto; min-width: 30px; width: 150px;"
                  >
                    <span
                      class="inline-flex items-center truncate whitespace-nowrap rounded-md font-medium ring-1 ring-inset px-2.5 py-1 text-sm text-xs bg-blue-50 text-blue-800 ring-blue-600/20"
                    >
                      Member
                    </span>
                  </td>
                  <td
                    class="text-sm text-gray-500 h-full px-4 py-3 first:pl-6 last:pr-6 whitespace-nowrap"
                    role="cell"
                    style="box-sizing: border-box; flex: 300 0 auto; min-width: 300px; width: 300px;"
                  >
                    <div
                      class="flex flex-wrap gap-1"
                    >
                      <span
                        class="inline-flex items-center truncate whitespace-nowrap rounded-md font-medium ring-1 ring-inset px-2.5 py-1 text-sm text-xs bg-gray-50 text-gray-800 ring-gray-600/20"
                      >
                        Test Workspace 2 Â· Viewer
                      </span>
                    </div>
                  </td>
                  <td
                    class="text-sm text-gray-500 h-full py-4 px-4 py-3 first:pl-6 last:pr-6 whitespace-nowrap"
                    role="cell"
                    style="box-sizing: border-box; flex: 150 0 auto; min-width: 30px; width: 150px;"
                  >
                    <time
                      class="whitespace-nowrap truncate"
                      datetime="2023-02-20T15:30:00.000+01:00"
                      title="2023-02-20T15:30:00.000+01:00"
                    >
                      February 20, 2023
                    </time>
                  </td>
                  <td
                    class="text-sm text-gray-500 h-full flex justify-end gap-x-2 px-4 py-3 first:pl-6 last:pr-6 whitespace-nowrap"
                    role="cell"
                    style="box-sizing: border-box; flex: 150 0 auto; min-width: 30px; width: 150px;"
                  >
                    <button
                      aria-label="edit"
                      class="relative h-fit
    inline-flex items-center justify-center transition-all
    border font-medium
    disabled:opacity-50 disabled:cursor-not-allowed
   rounded-sm focus:outline-hidden focus:ring-2 focus:ring-offset-2 text-xs px-2 py-1 shadow-xs border-transparent text-indigo-700 bg-indigo-100 hover:bg-indigo-200 focus:ring-indigo-500"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="h-4"
                        data-slot="icon"
                        fill="none"
                        stroke="currentColor"
                        stroke-width="1.5"
                        viewBox="0 0 24 24"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="m16.862 4.487 1.687-1.688a1.875 1.875 0 1 1 2.652 2.652L6.832 19.82a4.5 4.5 0 0 1-1.897 1.13l-2.685.8.8-2.685a4.5 4.5 0 0 1 1.13-1.897L16.863 4.487Zm0 0L19.5 7.125"
                          stroke-linecap="round"
                          stroke-linejoin="round"
                        />
                      </svg>
                    </button>
                    <button
                      aria-label="delete"
                      class="relative h-fit
    inline-flex items-center justify-center transition-all
    border font-medium
    disabled:opacity-50 disabled:cursor-not-allowed
   rounded-sm focus:outline-hidden focus:ring-2 focus:ring-offset-2 text-xs px-2 py-1 shadow-xs border-transparent text-indigo-700 bg-indigo-100 hover:bg-indigo-200 focus:ring-indigo-500"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="h-4"
                        data-slot="icon"
                        fill="none"
                        stroke="currentColor"
                        stroke-width="1.5"
                        viewBox="0 0 24 24"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="m14.74 9-.346 9m-4.788 0L9.26 9m9.968-3.21c.342.052.682.107 1.022.166m-1.022-.165L18.16 19.673a2.25 2.25 0 0 1-2.244 2.077H8.084a2.25 2.25 0 0 1-2.244-2.077L4.772 5.79m14.456 0a48.108 48.108 0 0 0-3.478-.397m-12 .562c.34-.059.68-.114 1.022-.165m0 0a48.11 48.11 0 0 1 3.478-.397m7.5 0v-.916c0-1.18-.91-2.164-2.09-2.201a51.964 51.964 0 0 0-3.32 0c-1.18.037-2.09 1.022-2.09 2.201v.916m7.5 0a48.667 48.667 0 0 0-7.5 0"
                          stroke-linecap="round"
                          stroke-linejoin="round"
                        />
                      </svg>
                    </button>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
          <div
            class="pointer-events-none absolute bottom-0 right-0 top-0 bg-linear-to-l to-transparent transition-opacity duration-100 from-white w-12 opacity-0"
          />
        </div>
      </div>
    </div>
  </article>
</div>
`;
